gh-121285: Remove backtracking when parsing tarfile headers

sethmlarson opened the pull request:
This removes all instances of backtracking from parsing tarfile headers, specifically `hdrcharset`, PAX, and GNU sparse headers.

<!-- gh-issue-number: gh-121285 -->
* Issue: gh-121285
<!-- /gh-issue-number -->

sethmlarson committed:
Remove backtracking when parsing tarfile headers

Eclips4 commented on Misc/NEWS.d/next/Security/2024-07-02-13-39-20.gh-issue-121285.hrl-yI.rst:
```diff
@@ -0,0 +1,2 @@
+Remove backtracking from tarfile header parsing for `hdrcharset`, PAX, and
```
```suggestion
Remove backtracking from tarfile header parsing for ``hdrcharset``, PAX, and
```

sethmlarson committed:
Update Misc/NEWS.d/next/Security/2024-07-02-13-39-20.gh-issue-121285.hrl-yI.rst

Co-authored-by: Kirill Podoprigora <kirill.bast9@mail.ru>

ethanfurman reviewed (changes requested):
Please remove the type hints.

sethmlarson committed:
Remove type hints

gpshead committed:
match up to 20 digits in _proc_gnusparse_00

this is used as an offset/length value, it doesn't need to be longer.  the \d+ prefix on the lines is not treated as an int so it's length bound is not important.

gpshead commented on Misc/NEWS.d/next/Security/2024-07-02-13-39-20.gh-issue-121285.hrl-yI.rst:
If the reporter wants to be credited, the news entry is also a common place to do so.

personnumber3377 replied:
I am the original reporter. Can you give credit to "Elias Myllymäki"?

serhiy-storchaka commented on Lib/tarfile.py:
```diff
@@ -1437,9 +1437,15 @@ def _proc_pax(self, tarfile):
+        ):
+            match = re.match(br"^\d{1,20} hdrcharset=([^\n]+)\n", buf[hdrcharset_offset:])
+            if match is not None:
+                pax_headers["hdrcharset"] = match.group(1).decode("utf-8")
```
You can simply use `br"\d hdrcharset=(.++)\n"` or `br"\d hdrcharset=(.+?)\n"`. Even `br"\d hdrcharset=(.+)\n"` has linear complexity.

gpshead replied:
The length prefix of hdrcharset is more than a single digit.  I suppose you are suggesting that since it is being ignored here we don't need to match all of its digits and can just use `\d`?

The non-backtracking (++ et al) feature is new in 3.11 so won't work for older security backports. Since we need to support those it is nice to keep the same code between releases. Which is one reason why this PR headed in the direction away from over-use of regular expressions for header parsing.

It is also a bug for us to not be validating the length value on a hdrcharset record... (in both the old code and this new code).  That could lead to unintended parsing consequences where we interpret a file different than other implementations would.

I think it would be better to restructure this parsing a bit. Instead of pre-searching for a possible hdrcharset= field within the header before the loop below, we should defer knowledge of and use of `encoding` until after the while loop extracting keyword=value pairs below is complete. Only then check for one being hdrcharset and use that as the encoding for `._decode_pax_field` to process all of the others.  That'd treat this as far less of a special case and focus pax header parsing in that one single loop.

The format specification appears to be: https://pubs.opengroup.org/onlinepubs/9699919799/utilities/pax.html

sethmlarson replied:
@gpshead I've implemented your suggestion (and added more framing tests) here: https://github.com/python/cpython/pull/121286/commits/39a419cbc6e345da202546af3af4d1be08aeb0a7 Please take a look and let me know what you think.

serhiy-storchaka commented on Lib/tarfile.py:
```diff
@@ -1454,14 +1460,14 @@ def _proc_pax(self, tarfile):
         # of the complete record including the length field itself and
         # the newline. keyword and value are both UTF-8 encoded strings.
-        regex = re.compile(br"(\d+) ([^=]+)=")
+        regex = re.compile(br"^(\d{1,20}) ([^=]+)=")
```
You can use possessive qualifiers `br"(\d++) ([^=]++)="`.

gpshead replied:
if sticking with over-reliance on regular expressions in 3.11 and later, yes, true (thanks for pointing out the modern feature!).  Though I don't think we should do that.  See other comments.

serhiy-storchaka replied:
Non-greedy qualifier has almost the same effect here -- removes backtracking.

We can use new feature in new versions and older features in older versions.

serhiy-storchaka commented on Lib/tarfile.py:
```diff
@@ -1454,14 +1460,14 @@ def _proc_pax(self, tarfile):
+        regex = re.compile(br"^(\d{1,20}) ([^=]+)=")
         pos = 0
-        while match := regex.match(buf, pos):
+        while match := regex.match(buf[pos:]):
```
Why is this change? It makes the complexity quadratic, because at every iteration you make a copy of `buf`.

gpshead replied:
I believe this was consequence of the `^` anchored regex as.`.match(buf, pos)` does not anchor the `^` to pos, but to the absolute start of buf.

regardless, it does have that downside.  I didn't dive into the `buf = tarfile.fileobj.read(self._block(self.size))` above, but expect that means the total amount of data is limited so the consequences of this copying would at least be constrained.

Regardless, this isn't great.  I think what I would prefer instead of a single regex to attempt to parse these headers is more to follow the intent of the format:

extract leading digits from buf@pos until a non-digit is encountered. error if that is not a space, error if the length is nonsense. error if the remaining data cannot satisfy length. partition the remaining data per the length on `=` - error if the remainder started or ended with the `=` or no `=` was found.

off the top of my head in this comment, the logic would probably look something like:

```python
# at module scope:
_length_prefix_re = re.compile(rb'(\d{1,20}) ')`

# within the while loop:
if not match := _length_prefix_re.match(buf, pos):
    raise InvalidHeaderError(...)
try:
    length = int(match.group(1))
except ValueError:
    raise InvalidHeaderError(...)
if length <= 3:
    raise InvalidHeaderError(...)
kv = buf[match.end(1) + 1:match.start(1) + length - 1]  # is this right?
keyword, equals, value = kv.partition(b'=')
if not keyword or equals != b'=' or ... ensure no invalid chars in keyword or value ...:
    raise InvalidHeaderError(...)
... decode_pax_field stuff ...  # but see other comment, delay this until after the loop
pax_headers[keyword] = value
...
```

serhiy-storchaka replied:
It's the opposite. `^` matches at the beginning of the string or after `\n` (if multiline flag is set). If you use `match()`, `^` has no effect if used in `match()` without starting index. So this change does not make sense.

Actually, the code may be simplified/optimized even more by using `finditer()`.

personnumber3377 replied:
The usage of the index as a parameter and splicing the buffer do act different:

```
import re
example_string = "\nabcdefg"
regex = re.compile("^abc")
ind = 1 # Some search index
# See different result:
print(regex.match(example_string, ind))
print(regex.match(example_string[ind:]))
```

the results from the print statements differ despite the regex having a carrot at the start. If the regex is changed to have a carrot at the start, then some tests will fail. With this buffer splicing all tests pass, but it is quite inefficient.

serhiy-storchaka commented on Lib/tarfile.py:
```diff
@@ -1520,12 +1526,26 @@ def _proc_pax(self, tarfile):
+            return values
+
+        offsets = finditer_without_backtracking(buf, b" GNU.sparse.offset=")
+        numbytes = finditer_without_backtracking(buf, b" GNU.sparse.numbytes=")
```
Just use `br"\d GNU.sparse.offset=(\d+)\n"` or `br"\d GNU.sparse.offset=(\d++)\n"`.

gpshead replied:
I suppose this is similar to above where the ignored probable length prefix doesn't need to be anchored and have its full digit values matched?  that makes this a bit more lenient than it already is such that it could possibly match things it should not in unobvious situations.  if the \d+ length prefix happens to be checked elsewhere as appears to be the case if this is only called by the earlier `._proc_pax` code that does that maybe it matters less. But what if someone encoded a value that matches these patterns but wasn't actually a header?  this blind regex approach already believes them even though they were not intended as a header and possibly malicious so that different software parsed them in a different manner...

serhiy-storchaka replied:
It matches the same things that before, but without backtracking.

sethmlarson replied:
I changed the implementation to use already-parsed headers: https://github.com/python/cpython/pull/121286/commits/4d04f3bc24f2f3e679722f8a06dee801bcff64ed Let me know what you think of this implementation.

gpshead reviewed (changes requested):
thanks for the additional reviews, this is turning into a deep dive on how tarfile parsing is implemented.  which it is probably a good thing to do. :)

fundamentally i believe this PR works and will resolve the original primary denial of service issues.  So we could go with it as is and backport it, overall things would be in better shape.

But I think there is further room for improvement to avoid misinterpreting some tar files in ways that may or may not be a security concern (if we parse them one way and other tar tools parse them another... that is a bug, but can also sometimes be a security problem).

Maybe we deal with that in followups?  Or do we tackle this all at once?

sethmlarson committed:
Rewrite PAX header parsing to be stricter

sethmlarson committed:
Optimize parsing of GNU extended sparse headers v0.0

sethmlarson committed:
Merge branch 'main' into remove-backtracking-tarfile-members

sethmlarson committed:
Merge branch 'main' into remove-backtracking-tarfile-members

sethmlarson committed:
Merge branch 'main' into remove-backtracking-tarfile-members

gpshead commented on Lib/tarfile.py:
```diff
@@ -1432,55 +1435,66 @@ def _proc_pax(self, tarfile):
+                length = int(match.group(1))
+            except ValueError:
+                raise InvalidHeaderError("invalid header")
+            # Headers must be at least 3 bytes, one for each of keyword, '=', and '\n'.
```
given the length field includes the size of the length field I think the minimum is 5?
`b"5 x=\n"`

per my reading of the comment above and https://pubs.opengroup.org/onlinepubs/9699919799/utilities/pax.html#tag_20_92_13_03

gpshead commented on Lib/tarfile.py:
```diff
@@ -1432,55 +1435,66 @@ def _proc_pax(self, tarfile):
+            if pos + length > len(buf):
+                raise InvalidHeaderError("invalid header")
+
+            keyword_and_value = buf[match.end(1) + 1:match.start(1) + length - 1]
```
`match.start(1) + length - 1` gets used twice. here, and again below to validate that the trailing `\n` 0x0A byte exists. assign it to a local.

gpshead commented on Lib/tarfile.py:
```diff
@@ -1432,55 +1435,66 @@ def _proc_pax(self, tarfile):
-            value = buf[match.end(2) + 1:match.start(1) + length - 1]
+            raw_headers.append((length, raw_keyword, raw_value))
+
+            # Check if the pax header contains a hdrcharset field. This tells us
```
What should we do if there are multiple hdrcharset= values in the extended pax header?  This loop combined with the one below will cause us to obey the first one that is set to BINARY, ignoring additional values for the PAX_NAME_FIELDS decoding that happens later. But the value ultimately exposed in our .pax_headers dict attribute will be that of the last one.

The https://pubs.opengroup.org/onlinepubs/9699919799/utilities/pax.html#tag_20_92_13_03 spec doesn't cover the case of multiple identically named fields being present. It makes sense to be an error to me... but matching what other tar/pax implementations do when reading a file with multiple identically named headers would be good to avoid files being interpreted differently by different tooling.

if nothing else, lets at least ensure we ignore future hdrcharset= fields within this header after we've used the value of one set to BINARY so that the value in our pax_headers dict winds up matching BINARY as that is how we behaved.

sethmlarson replied:
I addressed your comments in https://github.com/python/cpython/pull/121286/commits/6ca6329d1cca1dd25cd6ab8cf2a498bbe2a9154e, I went with the simpler approach of preserving previous behavior instead of doing a survey of existing tar tools. Hope this is okay for this fix?

sethmlarson committed:
Address review feedback

sethmlarson committed:
Fix invalid escape sequence
